### DISCUSSION TITLE
Improve workflow when force-pushing during code reviews

### DISCUSSION THREAD

**Discussion by sebastianst** [community/community#3478] [Pull Requests]
The [Github Flow article](https://guides.github.com/introduction/flow/) encourages pushing new commits while iterating over a pull requests during code reviews. However, this sometimes leads to a messy commit history and some teams prefer to continuously force-push clean commit histories rather than pushing little review commits.

This workflow doesn't work well with Github yet, unfortunately. [In 2018](https://github.blog/changelog/2018-11-15-force-push-timeline-event/), Github added a hyperlink to the word `force-pushed` in the conversation timeline to show the diff after a force-push, which is already a great improvement.

The following points are still missing in the Web UI to support smooth force-push review iterations:
- [ ] Add review comments to individual line(s) in the comparison view. Clicking on `force-pushed` shows the diff, but no line-comments can be made.
- [ ] Show all PR versions/heads in the comparison view, so that individual PR versions can be compared and reviewed.
- [ ] Show old line-comments on new force-pushed versions, if possible. This requires some sort of smart algorithm to match a line-comment to the new version.

Happy to add more points. Note that Gitlab and other platforms have better support for this kind of workflow for some time now.

Related issue: https://github.com/isaacs/github/issues/999

If you want to see these features being implemented, please vote by clicking on the :arrow_up: above the counter. Giving a :+1: _does not_ count as a vote, see also https://github.com/github/feedback/discussions/3474

**Comment by sebastianst** [comment]
Edit: Moved into "General Feedback" category, where it is well-placed. So the below discussion is obsolete.

Not sure if this was the right place to post this feature request [:point_right: Reposted to github.community](https://github.community/t/pull-requests-improve-workflow-when-force-pushing-during-code-reviews/166736).

**Reply by evi-liu** [reply]
Thank you for reposting to GitHub community! This space is currently reserved for GitHub Discussions feedback only.

**Reply by sebastianst** [reply]
Are there plans to open up github/feedback to other categories like the Pull Requests UI? I think the Discussions interface is nicer than github.community for discussions like these.

**Reply by aspiers** [reply]
See also https://github.com/isaacs/github/issues/1985

**Reply by sebastianst** [reply]
I moved this post into the "General Feedback" category, where it should be well placed. This makes my original comment obsolete.

**Comment by victorlin** [comment]
I would also like to see less clutter on the PR page when iterating on in-progress changes. Especially for draft PRs – I'm reluctant to open drafts with incomplete changes to avoid clutter on the PR page. This defeats the purpose of having a "draft" state.

In [Azure DevOps](https://docs.microsoft.com/en-us/azure/devops/repos/git/pull-requests), there is a similar amount of verbosity in commits/force-pushes, but it's not so bad because the PR page displays updates with most recent at the top. I prefer that over GitHub PRs where you have to scroll all the way down just to see the most recent changes/comments.

**Reply by hazz0003** [reply]
Azure DevOps some how has the same information density with about half the clutter.

**Comment by Empact** [comment]
Documented my prior experience with this in https://github.com/github/feedback/discussions/8808 - in my experience it's a significant source of friction between people who want to keep the review history and others who want to optimize for the source history.

**Comment by victorlin** [comment]
> Github added a hyperlink to the word `force-pushed` in the conversation timeline to show the diff after a force-push, which is already a great improvement.

I don't know if this was a recent change, but the designers somewhat addressed the verbosity issue here by compressing back-to-back force-pushes with a message _\<username\> force-pushed the \<branch\> branch n times, most recently from \<commit\> to \<commit\>_.

<img width="773" alt="image" src="https://user-images.githubusercontent.com/13424970/166336604-c22054c8-609f-428b-94f6-270373427a07.png">

However, the `force-pushed` link and **Compare** button only show the most recent force-push diff. There's no way to view the prior force-pushes, so this feels like an incomplete feature now.

**Reply by alpipego** [reply]
This should absolutely be expandable so each individual force push diff can be viewed. Moreover, often it would be more useful to show the diff of all the recent force pushes. I think the way the UI is done currently, it's impossible to get the individual hashes.

**Reply by danilaml** [reply]
Yes. It's actually a UX downgrade, since if one force-push was a rebase, and another was a fixup, it would combine their diffs into unreviewable mess, instead of comparing rebase with fixup.

**Reply by RalfJung** [reply]
Yeah, right now this "feature" means that it is impossible to reliably keep track of what changes in force-pushes. It's pretty bad.

Github might document "add new commits", but in my experience, a large fraction of contributors will use force-pushes anyway. It takes a lot of time an effort to teach them not to do that. It would be better if Github hat native support for standard git workflows such as force-push and rebase.

**Comment by alpipego** [comment]
One more thing that would be really helpful: The diff should not show rebased commits from the target branch of the PR.

If a feature branch gets rebased on the target branch and there are also code changes in the feature branch, the diff for the force pushes shows the changes of all commits from the target branch as well.

Since they have already been reviewed and merged, it's often very hard if not impossible for reviewers to identify the actual code changes of the push.

**Comment by Hnasar** [comment]
I sure hope this gets implemented one day! I think this is an area that GitLab does better than GitHub. 

`git` on the command line offers `range-diff` ([docs](https://git-scm.com/docs/git-range-diff)) which properly shows the diff between two sequences of commits and is a way that sensible changes can be seen, even in the face of force pushes and rebases.

This would be a killer code review feature for GitHub if it offered it in the web UI.

**Comment by RalfJung** [comment]
Another problem with the current situation is that when there were many force-pushes since a review, one has to look at all of them separately. It would be much better if Github kept track of the state (commit) of a PR when a review is submitted, and then offer a "show changes since this review" button that covers *all* the force-pushes (and other pushes) since then. Right now, I don't think there is even a manual way to get that commit, so I often have no choice but to re-review and entire PR, losing a lot of time.

**Comment by aspiers** [comment]
Unfortunately GitHub's PR flow of multiple commits per code review is fundamentally broken, and I doubt it will ever get properly fixed without changing the model.  This was already clear in 2017 when I first submitted https://github.com/isaacs/github/issues/999 (and also https://github.com/isaacs/github/issues/997 and https://github.com/isaacs/github/issues/998).  At that point I'd used both GitHub and Gerrit for 5+ years, and it was already very obvious how much better Gerrit's 1-commit-per-review model was, even before Gerrit enjoyed substantial improvements in UX etc.  So I fear this is flogging a dead horse.

One of the biggest flaws of this model is that GitHub has no stable unique id per commit in the PR, so it can't keep track of which commit is which during a force-push, since a force-push can add, remove, amend, and reorder commits arbitrarily.  If a new [patch-id](https://git-scm.com/docs/git-patch-id) appears, it's impossible to tell whether that is related to a previous commit.  Consequently the UI has no way to present the change from the previous version of the PR in an intelligible way.  Gerrit solves this by adding a stable unique id to every commit message as part of the review submission request, and it's a total game-changer.

This broken model leads to many additional problems, not least huge pains when trying to [stack PRs on top of other PRs](https://github.com/isaacs/github/issues/867).  [Many attempts have been made to solve this](https://github.com/orgs/community/discussions/4477#discussioncomment-1039627), but unfortunately they are all just bandaids.

The real solution is to limit PRs to 1 commit, use stable uuids per commit, use git commit's graph to natively express dependencies between reviews, and provide tooling which makes it quick and easy to submit a chain of _n_ commits as _n_ PRs.  In fact, it's possible to do this today using Gerrit with some extra GitHub compatibility plugins.  IIRC open source projects can even get [hosting for free from GerritHub](https://www.youtube.com/watch?v=jeWTvDad6VM).  The drawback is that it's not a well-trodden path, so things like GitHub Actions will require effort to integrate.

**Reply by RalfJung** [reply]
> Unfortunately GitHub's PR flow of multiple commits per code review is fundamentally broken

I don't agree with that. I never used Gerrit but did use Phabricator a little and found it an unpleasant experience compared to Github. (But maybe Gerrit is very different from Phabricator so not sure if this is a valid comparison.)

One-commit-per-review has its advantages, but it's not necessary to throw out the baby with the bathwater here. Github already has enough data to identify a proposed change in a given state (the pair of commits identifying the head of the branch in the PR, and the commit on master that it is based on); all we need is UI support for a range-diff between two such states.

**Reply by aspiers** [reply]
> Github already has enough data to identify a proposed change in a given state (the pair of commits identifying the head of the branch in the PR, and the commit on master that it is based on)

I don't see how this is sufficient.  If a rebase followed by force push amends a commit to slightly change its contents **and** reorders the sequence of commits, how can you link the new version of that commit with the old one?  Bearing in mind that neither the commit message nor the patch-id can be relied on as stable.

**Reply by mathstuf** [reply]
While I appreciate the work Gerrit does in this space, it still falls short as Commit-Id trailers aren't suitable in the face of splitting or merging patches either. IMO, getting commit message review support and per-commit diff views (GitLab has this one already) does a better job at spanning more of the problem space. I still like having a whole-branch view when reviewing as the end result is also of interest. It also allows for rebase-with-squashing-fixes-in-later workflows instead of requiring "perfect" rebases to properly allocate commit ids before pushing any content to the review system.

**Reply by danilaml** [reply]
@aspiers why would you need to link? The equivalent for gerrit "commit/patch-id" in GH would be the whole MR. You only need to be able to diff the MR contents before and after force-push excluding changes present in master/whatever branch MR is targeting.

**Reply by aspiers** [reply]
@mathstuf wrote:
> While I appreciate the work Gerrit does in this space, it still falls short as Commit-Id trailers aren't suitable in the face of splitting or merging patches either.

I wouldn't classify that as falling short.  It's impossible for any tool to perfectly _identify_ every change in version _n+1_ of a PR as either new or linked to an exact previous hunk in version _n_.  At some point you end up with unreliable heuristics and it's better to expect the humans to figure things out.  E.g. if a commit from version _n_ is split in version _n+1_, the author should be smart enough to decide which commit gets to keep the original Change-Id, or if it should be ditched because the split effectively amounts to a substantial reworking of the original.  Same with merging.  However Gerrit optimises for the frequent simpler cases where commits undergo relatively small amendments, and where it's useful to be able to link those versions together under a single review so that the reviewer can see the progression of the code through the review process.  **In particular**, this allows the crucial feature that they can easily verify whether their feedback has been correctly addressed.

> IMO, getting [...] per-commit diff views (GitLab has this one already) does a better job at spanning more of the problem space.

As per the previous point, it's necessary to be able to compare two revisions of the same commit.  Basic per-commit diff views don't allow this per se.

> I still like having a whole-branch view when reviewing as the end result is also of interest.

Sure, this is just tablestakes which every tool already offers.

> It also allows for rebase-with-squashing-fixes-in-later workflows instead of requiring "perfect" rebases to properly allocate commit ids before pushing any content to the review system.

I don't understand this point.  What allows (or even prevents) rebase with squashing?  Also, Change-Ids are allocated automatically on creation of each commit via a hook, so you don't need a rebase ("perfect" or otherwise) to allocate them.

**Reply by aspiers** [reply]
@danilaml wrote:
> @aspiers why would you need to link? The equivalent for gerrit "commit/patch-id" in GH would be the whole MR.

Since we're talking about GitHub not GitLab I assume you mean PR not MR?  Anyway, that is not an equivalent because any unique identifier for the PR only exists in GitHub, not in your local git metadata.  The things which link the local commit(s) to the PR are a) the local branch name, b) the PR branch name on GitHub, c) your local git configuration and usage of git which determines when pushing how to map from the local branch name to the remote name.  None of that helps when you have a sequence of 20 commits, you amend the 7th, and need the code review platform to figure out that the 7th commit has changed but none of the others really did.

> You only need to be able to diff the MR contents before and after force-push excluding changes present in master/whatever branch MR is targeting.

No, that's not all that's needed.  In the above example, if I'm the reviewer, I need to be able to:

- quickly and easily verify whether my feedback on the 7th commit was correctly addressed _without looking at any other changes_,
- not have to re-review _any_ of the other 19 commits which didn't change,
- for any of the other 19 commits which did change, be able to see exactly which ones did, and _how_ they changed.

If you think that can somehow be accomplished without marking each commit with a unique Change-Id or similar, please do share your proposal!  I'm 99% confident it's not possible, but totally open to being wrong.

**Reply by danilaml** [reply]
@aspiers You don't need to
>figure out that the 7th commit has changed but none of the others really did.

This is not what's being asked here. You only need to view the diff from state A and force-pushed state B (so you could tell if force-push was a 
simple rebase or actually contained some changes that need a new review). So you'd only see the changes between original and amended commit. That's enough to improve workflow dramatically and is currently possible to do offline via CLI.

**Reply by aspiers** [reply]
> This is not what's being asked here.

It's what _I'm_ asking for.  I posit that any review system incapable of doing this is fundamentally too limited.  If one only ever has to review small PRs which arrive infrequently enough not to have to worry about stacking, then OK this might be overkill.  But for any serious software project of size and change volume this is an absolute necessity.

> You only need to view the diff from state A and force-pushed state B

If by "state A" and "state B" you are referring to the tip of the PR branch, then I don't see how this satisfies any of the three requirements I stated above.  It sounds like you are assuming that

- there are only small changes between A and B (which can be completely invalidated by a rebase if the base of B is significantly newer than the base of A)
- these small changes can be easily attributed to amendments of specific commits within the 20 just by quick visual inspection
- B hasn't introduced commits which make substantial changes and then following commits which cancel out the majority of those

But none of these are guaranteed true, and even if they are, to be sure I would have to re-review 20 commits _every single force-push_, which is hopelessly inefficient.

>  (so you could tell if force-push was a simple rebase or actually contained some changes that need a new review)

How _exactly_ would you do this?  Again it sounds like you are relying on visual inspection, but when you are dealing with a PR with a large number of commits this is hopelessly inefficient.

> So you'd only see the changes between original and amended commit. That's enough to improve workflow dramatically and is currently possible to do offline via CLI.

I've tried this technique for a well over a decade at this point.  Yes for simple cases it can be sufficient, but as soon as things get complex it's totally inadequate.

**Reply by modesto15** [reply]
Helping

**Reply by danilaml** [reply]
>It's what I'm asking for.

Great. It's irrelevant, since it's not your issue and this discussion started with you disagreeing that GH has enough data already to implement what's being asked in the current issue, not some imaginary workflow you imagine.

>It sounds like you are assuming that

I don't assume any of that.

>these small changes can be easily attributed to amendments of specific commits within the 20 just by quick visual inspection

git blame

>I would have to re-review 20 commits every single force-push, which is hopelessly inefficient.

No. You review a single pull request as a whole. The number of commits is largely irrelevant. They could be all squashed into one for all we know. It's no different from gerrit. It just allows to split work into logically different chunks (or keep "address feedback"/"fix formatting" commits to show the evolution of the PR until they are all amended before merge).

**Reply by aspiers** [reply]
> > It's what I'm asking for.
> 
> Great. It's irrelevant, since it's not your issue

It's not yours either.  How is that relevant to anything?  This is a public forum :man_shrugging: 

> and this discussion started with you disagreeing that GH has enough data already to implement what's being asked in the current issue, not some imaginary workflow you imagine.

Actually no, it started in 2017 when I posted https://github.com/isaacs/github/issues/999 and other related issues, from which this discussion is a continuation (as referenced by the OP).

The post of my recent post was to explain why GitHub's PR model has fundamental limitations which means it's very unlikely that many of the issues associated with force-pushing will ever be fixed. 

> > these small changes can be easily attributed to amendments of specific commits within the 20 just by quick visual inspection
> 
> git blame
>
> > I would have to re-review 20 commits every single force-push, which is hopelessly inefficient.
> 
> No. You review a single pull request as a whole. The number of commits is largely irrelevant. They could be all squashed into one for all we know. It's no different from gerrit. It just allows to split work into logically different chunks (or keep "address feedback"/"fix formatting" commits to show the evolution of the PR until they are all amended before merge).

Sorry, but reviewing a single PR as a whole and doing squash merging is an _entirely_ different (and simpler) workflow which has all the limitations I outlined above.  If it works for you, great.  Indeed, [I already acknowledged that it can work OK in simple projects](https://github.com/orgs/community/discussions/3478#discussioncomment-11355212).  But please don't try to claim that means this workflow also works as a substitute for an entirely different workflow, which is used by many of the top FOSS projects globally, including Chromium, Linux, and git itself (which I have personally been involved with in the past).

This conversation is unfortunately going nowhere, so this will be my last reply to you.

**Reply by RalfJung** [reply]
> I don't see how this is sufficient. If a rebase followed by force push amends a commit to slightly change its contents and reorders the sequence of commits, how can you link the new version of that commit with the old one?

You compare the new version of *the PR* against the old one. This thread isn't about linking commits, it's about force-pushes to PRs.

> It's what I'm asking for. I posit that any review system incapable of doing this is fundamentally too limited.

Then you should start a new discussion. This one here is specifically about how force-pushes interact with the existing PR system. Saying that review should be per-commit is entirely off-topic.

Yes, this is a public forum, and in a public forum it is generally frowned upon to derail a discussion by bringing up a loosely related but clearly distinct topic.

The relation of this topic with https://github.com/isaacs/github/issues/999 is in the part where it says "if you amend PR branches via force-push, there is no way in the web UI to see the difference between the versions of the PR". A proper range-diff between the two PR states would easily solve that. Nothing here says or implies "PRs should only have a single commit". And in fact if PRs had a single commit, that alone would solve nothing! We'd still need range-diff to compare two versions of that commit. So a single-commit-per-PR policy is neither necessary nor sufficient for the problem at hand, and therefore entirely off-topic.

**Comment by koffes** [comment]
Any news on this issue?

It would be immensely valuable to e.g. highlight all diff since last review, and not just on a file-by-file basis.

**Comment by nodeax** [comment]
Gitlab has this implemented. Having this drastically improves reviews and also opens up possibilities for different workflows.

**Comment by akloss-terion** [comment]
YMMV, but I have observed in the last few weeks that GitHub's stale review dismissal is much improved. I've tried a few sorts of rebase/squash operations and it seems if the committed text is unchanged, GitHub no longer marks the review as stale.
